core/modules/replica/ServiceManagementJob.cc

```diff
@@ -0,0 +1,150 @@
+        unsigned int requestExpirationIvalSec,
+        Controller::Ptr const& controller,
+        string const& parentJobId,
+        Job::Options const& options)
+    :   Job(controller,
+            parentJobId,
+            requestName,
+            options),
+        _allWorkers(allWorkers),
+        _requestExpirationIvalSec(requestExpirationIvalSec) {
+}
+
+
+ServiceManagementJobResult const& ServiceManagementBaseJob::getResultData() const {
+
+    LOGS(_log, LOG_LVL_DEBUG, context() << __func__);
+
+    if (state() == State::FINISHED) return _resultData;
+
+    throw logic_error(
+            "ServiceManagementBaseJob::" + string(__func__) +
+            "  the method can't be called while the job hasn't finished");
+}
+
+
+void ServiceManagementBaseJob::startImpl(util::Lock const& lock) {
+
+    LOGS(_log, LOG_LVL_DEBUG, context() << __func__);
+
+    auto const workerNames = allWorkers() ?
+        controller()->serviceProvider()->config()->allWorkers() :
+        controller()->serviceProvider()->config()->workers();
+
+    for (auto&& worker: workerNames) {
+        _resultData.serviceState[worker] = ServiceState();
+        _resultData.workers[worker] = false;
+        _requests.push_back(submitRequest(worker));
+    }
+
+    // In case if no workers are present in the Configuration
+    // at this time.
+
+    if (_requests.size() == 0) finish(lock, ExtendedState::SUCCESS);
+}
+
+
+void ServiceManagementBaseJob::cancelImpl(util::Lock const& lock) {
+
+    LOGS(_log, LOG_LVL_DEBUG, context() << __func__);
+
+    for (auto&& ptr: _requests) ptr->cancel();
+    _requests.clear();
+    _numFinished = 0;
+}
+
+
+void ServiceManagementBaseJob::onRequestFinish(ServiceManagementRequestBase::Ptr const& request) {
+
+    LOGS(_log, LOG_LVL_DEBUG, context() << __func__
+         << "  worker=" << request->worker()
+         << " id=" << request->id()
+         << " type=" << request->type()
+         << " state=" << request->state2string());
+
+    if (state() == State::FINISHED) return;
+
+    util::Lock lock(_mtx, context() + string(__func__) + "[" + request->id() + "]");
+
+    if (state() == State::FINISHED) return;
+
+    // Update counters and object state if needed.
+    _numFinished++;
+    if (request->extendedState() == Request::ExtendedState::SUCCESS) {
+        _resultData.serviceState[request->worker()] = request->getServiceState();
+        _resultData.workers[request->worker()] = true;
+    }
+
+    // Check for the completion condition of the job
+    if (_requests.size() == _numFinished) {
+        size_t const numSucceeded = count_if(
+            _requests.begin(), _requests.end(),
+            [](ServiceManagementRequestBase::Ptr const& ptr) {
+                return ptr->extendedState() == Request::ExtendedState::SUCCESS;
+            }
+        );
+        finish(lock, numSucceeded == _numFinished ? ExtendedState::SUCCESS :
+                                                    ExtendedState::FAILED);
+    }
+}
```

Comment:
All of these onRequestFinish function look really similar and there are a lot of them. Defining a base request class you could get something like
```
void Job::onRequestFinish(JobRequest::Ptr const& request) {
    LOGS(_log, LOG_LVL_DEBUG, request->getLogStr());
    if (state() == State::FINISHED) return;
￼￼    util::Lock lock(_mtx, context() + string(__func__) + "[" + request->id() + "]");
￼    if (state() == State::FINISHED) return;
    virtSpecialHandlingFunc(request);
    finish(lock, success, successes, failures);
}

Reply:
This won't work in general because of the input parameter to the method which needs to be of a specific type (of a request). Otherwise the virtual method would have to do the `dynamic_cast` or `static_cast`. I'd like to avoid that if possible.